apf/apfAggregateNumbering.cc

```diff
@@ -0,0 +1,421 @@
+    //  sum since they include in their strides
+    //  all the nodes from the local aggregation scope
+    int strd = getScopeStride();
+    int frst = getLocalFirstDof();
+    int scp_offset = (int)(!(bool)frst) * (strd * dofs_per_blk * blks_per_nd);
+    int lcl_offset = PCU_Exscan_Int(scp_offset);
+    lcl_offset += scp_offset; // make the excan a scan
+    // remove the contribution from the local
+    //  aggregation scope, because the scope can be
+    //  any comm and only rank 0 in each scope
+    //  contributes to the above scan
+    lcl_offset -= (strd * dofs_per_blk * blks_per_nd);
+    frst += lcl_offset;
+    // strds and frsts needs to be resized/updated
+    //  since we can have adjacent parts that were
+    //  not in the aggregation scope under which
+    //  init was called
+    int sz = PCU_Comm_Peers();
+    frsts.setSize(sz);
+    MPI_Allgather(&frst,1,MPI_INTEGER,
+                  &frsts[0],1,MPI_INTEGER,
+                  gbl_cm);
+    // each aggregation scope has its own stride
+    //  we store per-rank since that is vastly
+    //  less complicated than keeping track
+    //  of the initial scopes
+    strds.setSize(sz);
+    MPI_Allgather(&strd,1,MPI_INTEGER,
+                  &strds[0],1,MPI_INTEGER,
+                  gbl_cm);
+  }
+  template <class T>
+  void AggregateNumberingOf<T>::getAll(MeshEntity * e, T * dat)
+  {
+    reinterpret_cast<FieldDataOf<T>*>(FieldBase::data)->get(e,dat);
+  }
+  template <class T>
+  T AggregateNumberingOf<T>::get(MeshEntity * e, int nd, int cmp)
+  {
+    int cmps = blks_per_nd * dofs_per_blk;
+    int nds = FieldBase::shape->countNodesOn(FieldBase::mesh->getType(e));
+    NewArray<T> data(nds*cmps);
+    getAll(e,&data[0]);
+    return data[nd*cmps + cmp];
+  }
+  // explicit instantiations
+  template class AggregateNumberingOf<int>;
+  template class AggregateNumberingOf<long>;
+  AggNumbering * createAggNumbering(Field * f,
+                                    int blocks,
+                                    int dofs_per_block,
+                                    MPI_Comm cm,
+                                    Sharing * share)
+  {
+    bool dlt = false;
+    if(!share)
+    {
+      share = getSharing(getMesh(f));
+      dlt = true;
+    }
+    AggNumbering * n = new AggNumbering;
+    MPI_Comm pcu_cm = PCU_Get_Comm();
+    bool swtch = pcu_cm != cm;
+    Sharing * init_share = share;
+    if(swtch)
+    {
+      init_share = new StaticSharing(getMesh(f),share,cm);
+      PCU_Switch_Comm(cm);
+    }
+    n->init(f,init_share,blocks,dofs_per_block);
+    if(swtch)
+    {
+      PCU_Switch_Comm(pcu_cm);
+      delete init_share;
+    }
+    // during init the static sharing may be used which
+    //  only provides binary ownership information, not
+    //  the owner rank of unowned entities, so we replace
+    //  it with the passed-in sharing here before globalization
+    //  which requires the rank information
+    n->setSharing(share);
```

Comment:
This may not be required anymore since StaticSharing now supplies ownership information as well.. but may also be the cleaner solution in the end